pf4j/src/main/java/org/pf4j/PluginClassLoader.java

```diff
@@ -107,137 +115,102 @@ public void addFile(File file) {
-                log.trace("Found resource '{}' in parent", name);
+                log.trace("Found resource '{}' in {} classpath", name,classLoadingSource);
                 return url;
+            } else {
+                log.trace("Couldn't find resource '{}' in {}.", name,classLoadingSource);
             }
-
-            log.trace("Couldn't find resource '{}' in parent", name);
-
-            url = findResourceFromDependencies(name);
-
-            if (url != null) {
-               log.trace("Found resource '{}' in dependencies", name);
-               return url;
-            }
-
-            return findResource(name);
         }
+        return null;
+
     }
 
     @Override
     public Enumeration<URL> getResources(String name) throws IOException {
-    	List<URL> resources = new ArrayList<>();
-
-    	if (!parentFirst) {
-            resources.addAll(Collections.list(findResources(name)));
-            resources.addAll(findResourcesFromDependencies(name));
-
-            if (getParent() != null) {
-                resources.addAll(Collections.list(getParent().getResources(name)));
-            }
-        } else {
-        	if (getParent() != null) {
-                resources.addAll(Collections.list(getParent().getResources(name)));
+        List<URL> resources = new ArrayList<>();
+
+        log.trace("Received request to load resources '{}'", name);
+        for (ClassLoadingStrategy.Source classLoadingSource : classLoadingStrategy.getSources()) {
+            switch (classLoadingSource) {
+                case PARENT:
+                    if (getParent() != null) {
+                        resources.addAll(Collections.list(getParent().getResources(name)));
+                    }
+                    break;
+                case PLUGIN:
+                    resources.addAll(Collections.list(findResources(name)));
+                    break;
+                case DEPENDENCIES:
+                    resources.addAll(findResourcesFromDependencies(name));
+                    break;
             }
-
-        	resources.addAll(findResourcesFromDependencies(name));
-        	resources.addAll(Collections.list(super.findResources(name)));
         }
+        return Collections.enumeration(resources);
 
-    	return Collections.enumeration(resources);
     }
 
-    private Class<?> loadClassFromDependencies(String className) {
+    protected Class<?> loadClassFromDependencies(String className) {
         log.trace("Search in dependencies for class '{}'", className);
         List<PluginDependency> dependencies = pluginDescriptor.getDependencies();
         for (PluginDependency dependency : dependencies) {
             ClassLoader classLoader = pluginManager.getPluginClassLoader(dependency.getPluginId());
 
             // If the dependency is marked as optional, its class loader might not be available.
-            if (classLoader == null && dependency.isOptional()) {
+            if (classLoader == null || dependency.isOptional()) {
```

Comment:
I think that is not good. You revert the bug solved by #386.

Reply:
oops , solved